src/bitcoin-chainstate.cpp

```diff
@@ -0,0 +1,255 @@
+
+const extern std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
+
+// Adapted from rpc/mining.cpp
+class submitblock_StateCatcher final : public CValidationInterface
+{
+public:
+    uint256 hash;
+    bool found;
+    BlockValidationState state;
+
+    explicit submitblock_StateCatcher(const uint256 &hashIn) : hash(hashIn), found(false), state() {}
+
+protected:
+    void BlockChecked(const CBlock& block, const BlockValidationState& stateIn) override {
+        if (block.GetHash() != hash)
+            return;
+        found = true;
+        state = stateIn;
+    }
+};
+
+
+int main(int argc, char* argv[]) {
+    assert(argc == 2);
+    std::filesystem::path abs_datadir = std::filesystem::absolute(argv[1]);
+    std::filesystem::create_directories(abs_datadir);
+    gArgs.ForceSetArg("-datadir", abs_datadir.string());
+
+    SelectParams(CBaseChainParams::MAIN);
+    const CChainParams& chainparams = Params();
+
+    init::SetGlobals();  // ECC_Start, etc.
+
+    // Initialize signatureCache cuz it's used by...
+    //     <- VerifyECDSASignature
+    //     <- CheckECDSASignature
+    //     <- EvalChecksigPreTapscript
+    //     <- EvalScript
+    //     <- VerifyScript
+    //     <- CScriptCheck::()
+    //     <- CheckInputScripts
+    //     <- ConnectBlock
+    //     <- ConnectTip
+    //     <- ActivateBestChainStep
+    //     <- ActivateBestChain
+    //     <- ProcessNewBlock
+    InitSignatureCache();
+
+    // Initialize g_scriptExecutionCache{,Hasher} cuz it's used by...
+    //     <- CheckInputScripts
+    //     <- ConnectBlock
+    //     <- ConnectTip
+    //     <- ActivateBestChainStep
+    //     <- ActivateBestChain
+    //     <- ProcessNewBlock
+    InitScriptExecutionCache();
+
+    // START scheduler for RegisterSharedValidationInterface
+    CScheduler scheduler{};
+    // Start the lightweight task scheduler thread
+    scheduler.m_service_thread = std::thread(util::TraceThread, "scheduler", [&] { scheduler.serviceQueue(); });
+
+    // Gather some entropy once per minute.
+    scheduler.scheduleEvery([]{
+        RandAddPeriodic();
+    }, std::chrono::minutes{1});
+    // END scheduler for RegisterSharedValidationInterface
+
+    GetMainSignals().RegisterBackgroundSignalScheduler(scheduler);
+
+    ChainstateManager chainman;
+
+    auto rv = LoadChainstate(false,
+                             std::ref(chainman),
+                             nullptr,
+                             false,
+                             chainparams.GetConsensus(),
+                             false,
+                             2 << 20,
+                             2 << 22,
+                             (450 << 20) - (2 << 20) - (2 << 22),
+                             false,
+                             false,
+                             [](){ return false; });
+    if (rv.has_value()) {
+        std::cerr << "Failed to load Chain state from your datadir." << std::endl;
+        goto epilogue;
```

Comment:
Wouldn't the code be easier to read by replacing this with `return;` and letting a destructor of an object that lives in this scope take care of the cleanup?

Reply:
Unfortunately, with our codebase right now if we don't do the shutdown process in a very carefully sequenced way, there will be nullptr dereferences and other undefined behaviour.

Reply:
> I don't mean to reorder the sequence. It is just a suggestion to remove `goto`, which in my eyes is impossible to review.

IMO, it is ok to use goto in a limited way to exit a block, and not worse than `break`. This can be cleaner and better than more convoluted alternatives and there are good practices build around it https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions

Reply:
I don't think we should be using the style guide of the kernel (a project written in C) for a project written in C++.

But given that the code here isn't used for anything after compilation (other than to prove that linking works), I think any style is fine here.

Reply:
Leaving this thread open for the benefit of future reviewers.

Reply:
Here's a simple way of avoiding goto without (I think) making things less readable: https://github.com/ajtowns/bitcoin/commit/5d7d8de713961a88a40c236346bbfad4ad2906f0

Reply:
@ajtowns Hmmm good point... Well since I'm going to introduce a `kernel::Context` object anyway, might be good to just move the init/shutdown logic over there and let RAII handle it. Will play around with it, thanks for the hint!!